connector_search_engine/components/mapper.py

```diff
@@ -20,8 +19,25 @@ def __init__(self, work):
         super(JsonExportMapper, self).__init__(work)
         exporter = work.index.exporter_id
         self._json_parser = exporter.get_json_parser()
-        if _f("id") not in self._json_parser["fields"]:
-            self._json_parser["fields"].append(_f("id"))
 
     def _apply(self, map_record, options=None):
-        return map_record._source.jsonify(self._json_parser)[0]
+        res = map_record._source.jsonify(self._json_parser)[0]
+
+        # Search engine mapper compatibility
+        #
+        # Some search engine like Algolia do not use the key "id" as key for the
+        # record but use a special one like "objectID"
+        # If the key defined globaly in _record_id_key is not in the result of
+        # the mapper and if the key "id" exist in the result
+        # then we automatically put the value of the "id" as value
+        # of the key _record_id_key
+        #
+        # example: for shopinvader all the exporter define the mapping for the key "id"
+        # then if you use Algolia there is no need of a glue module
+        # the "objectID" is will have the same value as the value of "id"
+        #
+        # If you need something different you can add a mapping in your exporter
+
+        if self.collection._record_id_key not in res and "id" in res:
```

Comment:
IMO this should not stay here but only in algolia implementation. Also, it should not add a new key but replace the new one.

NOTE that some time ago I moved [validation to backend type](https://github.com/OCA/search-engine/blob/14.0/connector_search_engine/models/se_backend_spec_abstract.py#L62) and by default it checks if the `_record_id_key` is there.

So, if you forget to provide `id` or `objectID` in json data you'll get an error.

Reply:
I think that it should add a new key to make the templating more compatible.

Indeed if you want to add a product in the cart whatever you use algolia or elastic you want to add "product.id" and not "product.objectID" for algolia and "product.id" for elastic

So in case of algolia you will have two key "id" and "objectID" that have the same value (one for the templating and one for the synchronisation between odoo and algolia)

What do you think?

Note the code is generic but for sure this "hack" is only need for algolia so we can keep it here or move in algolia and only move in the generic if an other search engine need the same behaviour

Reply:
Note for the validator it's still here: https://github.com/OCA/search-engine/blob/ff3576e636ae739f4acd578baf0c96950cd2c6ee/connector_algolia/components/adapter.py#L54

I didn't change the behaviour

Reply:
> I think that it should add a new key to make the templating more compatible.
> 
> Indeed if you want to add a product in the cart whatever you use algolia or elastic you want to add "product.id" and not "product.objectID" for algolia and "product.id" for elastic
> 
> So in case of algolia you will have two key "id" and "objectID" that have the same value (one for the templating and one for the synchronisation between odoo and algolia)
> 
> What do you think?
> 
> Note the code is generic but for sure this "hack" is only need for algolia so we can keep it here or move in algolia and only move in the generic if an other search engine need the same behaviour

hmm I see. Yeah I don't like it because of Algolia index's size constrains but I get your point. Then let's keep it but move it to Algolia. Fine?

Reply:
@simahawk done